src/java.compiler/share/classes/javax/lang/model/package-info.java

```diff
@@ -46,12 +46,85 @@
+
+ * <p>For example, consider the uses of {@code
+ * java.lang.String} in the string processing method {@code
+ * identityOrEmpty} below:
+ *
+ * {@snippet lang=java :
+ * // Return the argument if it is non-null and the empty string otherwise.
+ * public static @DefinitelyNotNull String identityOrEmpty(@MightBeNull String argument) {
+ *    ...
+ * }
+ * }
+ *
+ * The return type of the method is a {@code String} annotated with
+ * a {@code @DefinitelyNotNull} type annotation while the type of
+ * the parameter is a {@code String} annotated with a {@code
+ * @MightBeNull} type annotation. In a reflective API, since the set
+ * of annotations is different for the two <em>uses</em> of {@code
+ * String} as a type, the return type and argument type would need to
+ * be represented by different objects to distinguish between these two
+ * cases. The <em>definition</em> of {@code java.lang.String} itself
+ * is annotated with neither of the type annotations in question.
+ *
+ * <p>Another example, consider the declaration of the generic
+ * interface (JLS {@jls 9.1.2}) {@code java.util.Set} which has one
+ * type parameter. This declaration captures commonality between the
+ * many parameterized types (JLS {@jls 4.5}) derived from that
+ * declaration such as {@code java.util.Set<String>}, {@code
+ * java.util.Set<E>}, {@code java.util.Set<?>}, and also the raw type
+ * (JLS {@jls 4.8}) {@code java.util.Set}.
+ *
+ * <h3><a id=elementTypeMapping>Mapping between Elements and Types</a></h3>
+ *
+ * While distinct concepts, there are bidirectional (partial) mappings
+ * between elements and types, between definitions and uses. For
+ * example, roughly speaking, information that would be invariant for
+ * all uses of a type can be retrieved from the element defining a
+ * type. For example, consider a {@link
+ * javax.lang.model.type.DeclaredType DeclaredType} type mirror
+ * modeling a use of {@code java.lang.String}. Calling {@link
+ * javax.lang.model.type.DeclaredType#asElement()} would return the
+ * {@link javax.lang.model.element.TypeElement} for {@code
+ * java.lang.String}. From the {@code TypeElement}, common information
+ * such as {@linkplain
+ * javax.lang.model.element.TypeElement#getSimpleName() name} and
+ * {@linkplain javax.lang.model.element.TypeElement#getModifiers()
+ * modifiers} can be retrieved.
+ *
+ * <p>All elements can be {@linkplain
+ * javax.lang.model.element.Element#asType() mapped to} some type. For
+ * classes and interfaces, their elements get {@linkplain
+ * javax.lang.model.element.TypeElement#asType() mapped to} a
+ * prototypical type. Conversely, in general, many types can map to a
+ * given type element. For example, the type mirror for the raw type
+ * {@code java.util.Set}, the prototypical type {@code
+ * java.util.Set<E>}, and the type {@code java.util.Set<String>} would
+ * all {@linkplain javax.lang.model.type.DeclaredType#asElement() map
+ * to} the element for {@code java.util.Set}. Several kinds of types
```

Comment:
It would be good to mention at least one `e.g.` for each case - e.g. `DeclaredType` have an element, but `ExecutableType` does not.

Reply:
Updated per your suggestions, I'll reflow the paragraphs once we've settled on the final text.